### DISCUSSION TITLE
Fan wildcard not called

### DISCUSSION THREAD

**Discussion by freedreamer82** [RavenSystem/esp-homekit-devices#2043] [MEPLHAA Scripts]
Hi,

I'm trying to build a fan controller to 0-10V DC. I  would like to use an ESP32 and a PWM to 0-10V device.
the device has 2 fan.

Tested with:
ESP32 azdelivery with  FW 12.4.4 Merlin

```

 {
  "c":{"io": [   [[13],2],[[10],2] , [[9],2] , [[ 22,23 ], 7] ] },
  "a":[{
	"t":65,
	"e":4,
	"y0":[
		{"v": 0,  "0":{ "q": [ [ 22, 0, 2000 ] ]  } },
		{"v":25,  "0":{ "q": [ [ 22, 16384, 2000 ] ]  } },
		{"v":50,  "0":{ "q": [ [ 22, 32768, 2000 ] ]  } },
		{"v":75,  "0":{ "q": [ [ 22, 49152, 2000 ] ]  } } ,
	 	{"v":100, "0":{ "q": [ [ 22, 65535, 2000 ] ]  } }
    ],
    "es":[{
        "t":65,
	"e":4,
	"y0":[
		{"v": 0,  "0":{ "q": [ [ 23, 0, 2000 ] ]  } },
		{"v":25,  "0":{ "q": [ [ 23, 16384, 2000 ] ]  } },
		{"v":50,  "0":{ "q": [ [ 23, 32768, 2000 ] ]  } },
		{"v":75,  "0":{ "q": [ [ 23, 49152, 2000 ] ]  } } ,
	 	{"v":100, "0":{ "q": [ [ 23, 65535, 2000 ] ]  } }
     ]
  }],
}]
}
```

the above example is a first idea....I got from the Home App the 2 fan with 4 speed steps (0-25-50-100) but the wildcard will not be called.  I tried adding a GPIO with a led connected but with no luck 
@RavenSystem any idea? is something wrong with the script or with the FW ?

Another question :
If I would like to ADD 2 buttons per fan for a local handling , with the purpose of increase/decrease the fan speed,  should be possible? I didn't find anything related to this possibility.
The idea should be every press to increase the speed for a button and decrease for the other...how can I do ?

Thanks as usual

Marco

**Comment by RavenSystem** [comment]
Using `"e":4` will give you wildcard values of 1,2,3 and 4, not 25, 50, 75, and 100. https://github.com/RavenSystem/esp-homekit-devices/wiki/Fan#wildcard-actions

You can use buttons to increase/decrease speed, but you must declare them as switch services (outside HomeKit with "h":0), and then use Fan Service Notifications: https://github.com/RavenSystem/esp-homekit-devices/wiki/Fan#service-notifications

**Comment by freedreamer82** [comment]
ok , changed to 0,1,2,3 it works.

the idea to use a switch outside homekit seems a good one..but when i send notification , is possiblle to set the delta? I mean..if I send -25 it will decrease by 25 the actual value?

thanks

M.

**Reply by RavenSystem** [reply]
Yes, but to decrease speed by 25 steps, you must send a value of -125.

**Reply by freedreamer82** [reply]
but if i add a binary switch I have to increase both on "0" and "1" action , right? (if I want every press an increase.)

btw i guess i found a bug in the companion app.

if I hide the switch with h:0 in the app i don't see anymore the device and it is moved to "others" section. If i remove the hidden device , it works.

```
{
  "c":{"io": [   [[13],2],[[10],2] , [[9],2] ,[[12],6], [[14],6],[[ 22,23 ], 7] ] ,"b":[[0,5]] },
  "a":[{
	"t":65,
	"e":4,
	"0":{  "r":[ [9,0],[10,0] ,[13,0] ] },
	"y0":[
		{"v":1,  "0":{ "r" :[  [13,1],[9,0],[10,0] ]    } },
		{"v":2,  "0":{ "r" :[  [9,1],[13,0],[10,0] ]    } },
		{"v":3,  "0":{ "r" :[  [10,1],[9,0],[13,0] ]  } } ,
	 	{"v":4, "0":{  "r" :[  [10,1],[9,1],[13,1] ]   } }
    ],
    "es":[{
        "t":65,
	"e":4,
	"y0":[
		{"v": 0,  "0":{ "q": [ [ 23, 0, 2000 ] ]  } },
		{"v":1,  "0":{ "q": [ [ 23, 16384, 2000 ] ]  } },
		{"v":2,  "0":{ "q": [ [ 23, 32768, 2000 ] ]  } },
		{"v":3,  "0":{ "q": [ [ 23, 49152, 2000 ] ]  } } ,
	 	{"v":4, "0":{ "q": [ [ 23, 65535, 2000 ] ]  } }
     ]
  },
  {
    "t": 1,
    "0": { "m": [ [ 1, -75 ] ] },
    "1": { "m": [ [ 1, -75 ] ] },
    "b": [ [ 12 ] ],
    "h":0
  },
    {
    "t": 1,
    "0": { "m": [ [ 1, -125 ] ] },
    "1": { "m": [ [ 1, -125 ] ] },
    "b": [ [ 14 ] ],
    "h":0
  }
  
  ],
}]
}
```

PS
i have two fans but the first one is a test with LEDs

**Reply by RavenSystem** [reply]
"es":[] array can not have services without HomeKit ("h":0): https://github.com/RavenSystem/esp-homekit-devices/wiki/Accessory-Configuration#Extra-Services

**Reply by freedreamer82** [reply]
oh ... sorry , there a lot of combinations I always forget something... tnx!

**Reply by xrust83** [reply]
```{
  "c": {
    "io":[[[2,4,5,12,15],2],[[0,10,13,14],6]],
    "n":"WildWind-Fan",
    "l":2,
    "b":[[0,5]]
  },
  "a": [
    {
      "0": {
        "r":[[4],[5],[12]]
      },
      "t":65,
      "y0":[
        {"0":{"r":[[12,1],[4],[5]]},"v":1},
        {"0":{"r":[[4,1],[5],[12]]},"v":2},
        {"0":{"r":[[5,1],[4],[12]]},"v":3}
      ],
      "e":3,
      "es": [
        {
          "0":{"m":[[4,-20001],[5,-20001],[6,-20001]]},
          "1":{"m":[[4,-20000],[5,-20000],[6,-20000]]}
        },
        {
          "0":{"r":[[15]]},
          "1":{"r":[[15,1]]},
          "b":[[9]]
        }
      ]
    },
    {
      "0":{"m":[[1,101],[1,1]]},
      "1":{"m":[[1,101],[1,1]]},
      "b":[[10,0]],
      "h":0
    },
    {
      "0":{"m":[[1,101],[1,2]]},
      "1":{"m":[[1,101],[1,2]]},
      "b":[[13,0]],
      "h":0
    },
    {
      "0":{"m":[[1,101],[1,3]]},
      "1":{"m":[[1,101],[1,3]]},
      "b":[[14,0]],
      "h":0
    },
    {
      "0":{"m":[[1,0]]},
      "1":{"m":[[1,0]]},
      "b":[[0,0]],
      "h":0
    }
  ]
}
```

https://www.youtube.com/watch?v=IOK_PRNA5lA
